Move Fit2Obs to stand-alone job

KateFriedman-NOAA opened the pull request:
**Description**

This PR moves the Fit2Obs invocation out of the vrfy job and into its own dedicated `fit2obs` job in the gdas suite. This new dedicated job uses the latest Fit2Obs tag `wflow.1.0`.

The Fit2Obs requires a type of spin-up. The job looks back `VBACKUP_FITS` hrs and needs available inputs for that lookback cycle in the `ROTDIR`. The `jobs/JGDAS_FIT2OBS` script will first check that `xdate` (`CDATE` - `VBACKUP_FITS`) > `SDATE` and then if met, will check that the needed inputs exist.
- If `xdate>SDATE` is not yet satisfied, the job will exit 0 with "Too early for FIT2OBS to run. Exiting.". The conditional is greater-than and not greater-than-or-equal since the first half cycle generally does not have some of the needed inputs (e.g. prepbufr). Thus the first half cycle is not included in the valid lookback cycles. This avoids erroneous job failures for the first cycle to run the Fit2Obs package. Additional logic could be introduced to include the half cycle if all available inputs are available.
- If any of the needed inputs are missing the job will abort with "FATAL ERROR: FILE MISSING: ${file}". 

This spin-up means that the first cycles will run the job but exit 0 immediately. The 6th cycle (if `VBACKUP_FITS=24`) will be the first cycle to run the Fit2Obs package and produce output in the online archive.

Changes:

1. Remove fit2obs variables and settings from `config.vrfy` and into newly created `config.fit2obs` for `fit2obs` job.
2. Remove fit2obs submission/invocation from `jobs/rocoto/vrfy.sh`.
3. Create new `fit2obs` job scripts: `jobs/rocoto/fit2obs.sh` and `jobs/JGDAS_FIT2OBS`
4. Add new `fit2obs` job to setup scripts: `workflow/applications.py` and `workflow/rocoto/workflow_tasks.py`
5. Add new `fit2obs` job to all env files.
6. Add new `fit2obs` job into `config.resources` (use 1 node on WCOSS2 and 3 nodes elsewhere).
7. Add `export DO_FIT2OBS="YES"` to `config.base.emc.dyn`.

Resolves #1405
Resolves #1232
Resolves #1058

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

- [x] Clone and build tests on all supported platforms
- [x] Cycled tests on Hera, Orion, WCOSS2, and Jet (Jet tests done by @DavidHuber-NOAA )
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

KateFriedman-NOAA committed:
Fit2Obs updates for vrfy.sh and config.vrfy

- Add new variables to vrfy.sh (as done in v16 package).
- Update COM_PRP path variable to use "obs" instead of "atmos"
after recent change to ROTDIR structure for observations.
- Change "RUN_ENVIR" variable to "OUTPUT_FILETYPE" to avoid
confusion with how RUN_ENVIR is used (nco vs emc) and use
correct variable for defining type as "netcdf".
- Update fit_ver to tentative new version (update again if needed).
- Introduce some shellcheck updates to variables.
- Update PREPQFITSH script names for each platform.

Refs #1232

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Remove fit2obs from vrfy job

- Remove the fit2obs section from jobs/rocoto/vrfy.sh
- Remove fit2obs variables/settings from parm/config/config.vrfy

Refs #1232, #1405

KateFriedman-NOAA committed:
Add DO_FIT2OBS switch to config.base

Refs #1405, #1232

KateFriedman-NOAA committed:
Add fit2obs job to config.resources

- Add "fit2obs" to list of jobs in config.resources
- Add fit2obs resource block and values in config.resources

Refs #1405, #1232

KateFriedman-NOAA committed:
Add new fit2obs job to workflow setup scripts

- Add do_fit2obs switch.
- Add fit2obs to config list.
- Add fit2obs to task list.
- Create fit2obs task for gdas CDUMP.
- Add fit2obs as task dependency for arch job when do_fit2obs is true and CDUMP=gdas.

Refs #1405, #1232

KateFriedman-NOAA committed:
Create config.fit2obs for new fit2obs job

Refs #1405, #1232

KateFriedman-NOAA committed:
Create new rocoto job script for fit2obs job

- Create fit2obs.sh.
- Script will run new jobs/JGDAS_FIT2OBS script.

Refs #1405, #1232

KateFriedman-NOAA committed:
Create JJOB script for new fit2obs job

- Create jobs/JGDAS_FIT2OBS.
- Script will run fit2obs package batrun/excfs_gdas_vrfyfits.sh.ecf script.

Refs #1405, #1232

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Add fit2obs job to env files

Refs #1405

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Add fit2obs to gdas_half but don't run until xdate>=SDATE

- Add fit2obs job to first half cycle to resolve gdasarch dependency.
- Update JGDAS_FIT2OBS to check xdate (look-back cycle for fit2obs) against
SDATE and only run when xdate>=SDATE.
- Add load of netcdf/4.7.0 (needed before build is updated).

Refs #1405

KateFriedman-NOAA committed:
Update path settings in config.fit2obs and remove CDUMPFCST

- Remove CDUMPFCST, it gets set in fit2obs package.
- Update EXECcfs and USHcfs to include "install" in paths

Refs #1405

KateFriedman-NOAA committed:
Updates to JGDAS_FIT2OBS for functioning

- Add second CDATE setting (CDATE=${xdate}) for lookback date.
The fit2obs package uses CDATE and needs it set from above based
on the xdate variable (lookback date).
- Update COM_PRP to evaluate it up front.
- Add PRPI, sig1, sfc1, and CNVS input file variables for dependency
checking later in script.
- Add err_chk and postmsg setup (as done in runfits in fit2obs package).
- Add check of xdate against SDATE (-gt) for sufficient spinup of test.
- Add check of input file existence and exit with FATAL ERROR is any
are missing.
- Remove temporary "module load netcdf/4.7.0" now that build is updated.
- Indent section of script that's now within xdate/SDATE and input file
check. Only run fit2obs package script if dependencies are met.

Refs #1405

KateFriedman-NOAA committed:
Remove "/install" from EXECcfs and USHcfs paths

Refs #1405

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Update config.fit2obs for new wflow.1.0 Fit2Obs tag

- Update fit_ver="wflow.1.0".
- Update fitdir path to new install location on all supported platforms.

Refs #1405

aerorahul commented on env/HERA.env:
```diff
@@ -291,4 +292,14 @@ elif [[ "${step}" = "gempak" ]]; then
+
+    export NTHREADS_FIT2OBS=${nth_fit2obs:-1}
+    [[ ${NTHREADS_FIT2OBS} -gt ${nth_max} ]] && export NTHREADS_FIT2OBS=${nth_max}
+    export MPIRUN="${launcher} -n ${npe_fit2obs}"
```
too bad they use `MPIRUN` instead of `APRUN`

KateFriedman-NOAA replied:
Perhaps something we can change in a later Fit2Obs PR?

aerorahul commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+##########################################
+# Remove the Temporary working directory
+##########################################
+cd "${DATAROOT}" || (echo "${DATAROOT} does not exist. ABORT!"; exit 1)
```
```suggestion
cd "${DATAROOT}" || (echo "FATAL ERROR: ${DATAROOT} does not exist. ABORT!"; exit 1)
```

KateFriedman-NOAA replied:
The `jobs/JGDAS_ENKF_ARCHIVE` and `jobs/JGLOBAL_ARCHIVE` scripts should also have a similar update. I will commit the suggestion into `JGDAS_FIT2OBS` though.

KateFriedman-NOAA commented:
Working to resolve shellcheck warnings...

aerorahul commented on parm/config/config.fit2obs:
```diff
@@ -0,0 +1,30 @@
+export fit_ver="wflow.1.0"
+export fitdir="${BASE_GIT}/Fit2Obs/${fit_ver}"
+
+export HOMEcfs=${fitdir}
+export EXECcfs=${HOMEcfs}/exec
+export USHcfs=${HOMEcfs}/ush
```
Are these even used?

KateFriedman-NOAA replied:
They are used within the Fit2Obs package.

aerorahul replied:
ok. These may need a new definition as fit2obs conforms to EE2.  But for a later PR.

jack-woollen replied:
I'll work on changing the fit2obs package to conform in the not too distant future.

aerorahul replied:
for now, I can define these in the fit2obs modulefile so they don't need to appear here.

WalterKolczynski-NOAA commented on jobs/JGDAS_FIT2OBS:
A lot of this should be pushed down out of the j-job to an exscript.

KateFriedman-NOAA replied:
I'm treating the fit2obs job somewhat similarly to how we run the obsproc/prepobs package and invoking the external package driver script (`"${fitdir}/batrun/excfs_gdas_vrfyfits.sh.ecf"`) from the JJOB script. For the prep job we invoke the JJOB from that package from our `prep.sh` but there isn't a JJOB for fit2obs package so I created one and included the settings we need before invoking `excfs_gdas_vrfyfits.sh.ecf`. I didn't see a need to add another exscript and making the JJOB essentially do nothing but invoke it. Let me know if I'm missing a greater point though.

aerorahul commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+export CDATE=${CDATE:-${PDY}${cyc}}
+export CDUMP=${CDUMP:-${RUN:-"gdas"}}
+
+export xdate=$(${NDATE} -${VBACKUP_FITS} ${CDATE})
```
```suggestion
export xdate=$(${NDATE} -${VBACKUP_FITS} ${PDY}${cyc})
```

WalterKolczynski-NOAA replied:
```suggestion
export xdate=$(${NDATE} -"${VBACKUP_FITS}" "${PDY}${cyc}")
```

WalterKolczynski-NOAA commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+
+export xdate=$(${NDATE} -${VBACKUP_FITS} ${CDATE})
+export CDATE=${xdate} # Reset CDATE to xdate for lookback
+export vday=$(echo ${xdate} | cut -c1-8)
+export vcyc=$(echo ${xdate} | cut -c9-10)
```
```suggestion
export vday=${xdate:0:8}
export vcyc=${xdate:8:2}
```

KateFriedman-NOAA replied:
Committed change, thanks @WalterKolczynski-NOAA !

aerorahul commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+# Set variables used in the script
+##############################################
+
+export CDATE=${CDATE:-${PDY}${cyc}}
```
```suggestion
```

WalterKolczynski-NOAA replied:
```suggestion
```

aerorahul commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+export CDUMP=${CDUMP:-${RUN:-"gdas"}}
+
+export xdate=$(${NDATE} -${VBACKUP_FITS} ${CDATE})
+export CDATE=${xdate} # Reset CDATE to xdate for lookback
```
```suggestion
export CDATE=${xdate} # set CDATE to xdate for use in fit2obs package
```

WalterKolczynski-NOAA commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+export CDATE=${CDATE:-${PDY}${cyc}}
+export CDUMP=${CDUMP:-${RUN:-"gdas"}}
+
+export xdate=$(${NDATE} -${VBACKUP_FITS} ${CDATE})
+export CDATE=${xdate} # Reset CDATE to xdate for lookback
```
Is the external package using `$CDATE` or `$xdate`? Or neither? Don't export things that aren't needed downstream.

KateFriedman-NOAA replied:
The Fit2Obs package uses `CDATE` throughout whereas `xdate` is just used at the workflow level to reset `CDATE` to the lookback cycle and to set `vday` and `vcyc` before invoking the Fit2Obs scripts. A lot of this is lifted from how we were doing it in the vrfy job, it can be rethought if needed.

WalterKolczynski-NOAA replied:
I would just replace xdate with CDATE then to avoid the duplicate variable, although I don't like fit2obs using CDATE rather than VDATE.

KateFriedman-NOAA replied:
Good call, done at 3b48ae11. Thanks!

aerorahul commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+
+export PRPI=${COM_PRP}/${CDUMP}.t${vcyc}z.prepbufr
+export sig1=${COM_INA}/${CDUMP}.t${vcyc}z.atmanl.nc
+export sfc1=${COM_INA}/${CDUMP}.t${vcyc}z.atmanl.nc
```
Is this a mistake?
```suggestion
export sfc1=${COM_INA}/${CDUMP}.t${vcyc}z.sfcanl.nc
```

KateFriedman-NOAA replied:
This was lifted right from the Fit2Obs scripts: https://github.com/NOAA-EMC/Fit2Obs/blob/develop/batrun/excfs_gdas_vrfyfits.sh.ecf#L64
It may be a mistake in Fit2Obs package. @jack-woollen should the `sfc1` file be the `sfcanl.nc` file instead of the `atmanl.nc` file?

jack-woollen replied:
@KateFriedman-NOAA Well its not exactly correct, but a do no harm workaround. I need to write code to process the netcdf sfcanl, not a high priority at the moment.

WalterKolczynski-NOAA commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+##############################################
+
+export CDATE=${CDATE:-${PDY}${cyc}}
+export CDUMP=${CDUMP:-${RUN:-"gdas"}}
```
```suggestion
export CDUMP=${RUN/enkf}
```

WalterKolczynski-NOAA replied:
Also: is `$CDUMP` even needed?

KateFriedman-NOAA replied:
Yes, the Fit2Obs package uses `CDATE` throughout so we have to set it from above.

WalterKolczynski-NOAA replied:
Do you mean `CDUMP`? If they are using it for paths where they should be using `${RUN}`, that could cause problems if they are looking at ensemble members.

aerorahul commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+export vcyc=$(echo ${xdate} | cut -c9-10)
+
+export COM_INA=${ROTDIR}/gdas.${vday}/${vcyc}/atmos
+export COM_INF='$ROTDIR/vrfyarch/gfs.$fdy/$fzz'
```
I notice this is in `'` quotes, so `$fdy` and `$fzz` won't get resolved here.
Where are these used?

KateFriedman-NOAA replied:
The `COM_INF` needs to be passed into the Fit2Obs package with single quotes and then be evaluated there when `fdy` and `fzz` are set. I tried setting this with double quotes (based on shellcheck warning) but then it tries to evaluate it at the workflow level which fails since `fdy` and `fzz` aren't known yet.

WalterKolczynski-NOAA replied:
Made a suggestion to suppress the shellcheck warning

KateFriedman-NOAA replied:
Works for me, thanks @WalterKolczynski-NOAA !

WalterKolczynski-NOAA commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+export COMLOX=${DATA}/fitx
+[[ ! -d "${COMLOX}" ]] && mkdir -p "${COMLOX}"
+
+echo "echo err_chk">${DATA}/err_chk; chmod 755 ${DATA}/err_chk
+echo "echo postmsg">${DATA}/postmsg; chmod 755 ${DATA}/postmsg
```
Why?

KateFriedman-NOAA replied:
I took this from the Fit2Obs subfits scripts, which we now go around. I ran without these two lines and it fails. something in the fit2obs scripts needs this. Can explore in a follow-up PR how to do this better in fit2obs package and remove these two lines.

aerorahul commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,83 @@
+    cat "${pgmout}"
+  fi
+
+else
+
+  echo "Too early for FIT2OBS to run. Exiting."
+
```
I would flip this other way around and right up front after `xdate` is calculated like so:

```bash
if [[ "${xdate}" -le "${SDATE:-}" ]]; then
    echo "Too early for FIT2OBS to run. Exiting."
fi
``` 

And then here, you can remove the big `if-block`

KateFriedman-NOAA replied:
This would need an `exit 0` too, right? If we exit right here then the `DATAROOT` that is created by `jjob_header.sh` at the start of the script won't be removed at the end of `JGDAS_FIT2OBS`. That's why I set up the if-block like I did.

aerorahul reviewed (changes requested):
Some comments.  Some suggestions, and a couple questions.
Looks good otherwise.

WalterKolczynski-NOAA commented on parm/config/config.resources:
```diff
@@ -623,6 +623,14 @@ elif [ ${step} = "vrfy" ]; then
+    export npe_fit2obs=3
+    export nth_fit2obs=1
+    export npe_node_fit2obs=1
+    if [[ ${machine} == "WCOSS2" ]]; then export npe_node_fit2obs=3 ; fi
```
Why does WCOSS require three full nodes vs only 3 CPU on others?

KateFriedman-NOAA replied:
WCOSS2 only uses 1 node (npe_fit2obs / npe_node_fit2obs = 3/3 = 1), whereas the R&Ds use 3 nodes (3/1 =3). This is to get enough memory.

WalterKolczynski-NOAA replied:
I always forget it is defined backwards. When we rewrite the resources, we need to improve the variable name.

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -1020,6 +1034,9 @@ def arch(self):
         if self.app_config.do_vrfy:
             dep_dict = {'type': 'task', 'name': f'{self.cdump}vrfy'}
             deps.append(rocoto.add_dependency(dep_dict))
+        if self.app_config.do_fit2obs and self.cdump in ['gdas']:
+            dep_dict = {'type': 'task', 'name': f'{self.cdump}fit2obs'}
+            deps.append(rocoto.add_dependency(dep_dict))
```
Does metp actually depend on fit2obs?

KateFriedman-NOAA replied:
This is me adding fit2obs to the archive job dependency, not metp.

WalterKolczynski-NOAA replied:
d'oh, missed the code break where GH skipped lines

KateFriedman-NOAA committed:
Address SC2250 and SC2292 shellcheck in config.fit2obs

Refs #1405

KateFriedman-NOAA committed:
Address SC2086 and SC2292 shellcheck in JGDAS_FIT2OBS

Refs #1405

KateFriedman-NOAA committed:
Update jobs/JGDAS_FIT2OBS

Remove unneeded CDATE line.

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

KateFriedman-NOAA committed:
Update jobs/JGDAS_FIT2OBS

Adjust CDUMP setting.

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

WalterKolczynski-NOAA commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,82 @@
+export COM_INF="$ROTDIR/vrfyarch/gfs.$fdy/$fzz"
+export COM_PRP=${ROTDIR}/gdas.${vday}/${vcyc}/obs
+
+export PRPI=${COM_PRP}/${CDUMP}.t${vcyc}z.prepbufr
```
All these file names should use `${RUN}`, not `${CDUMP}`

KateFriedman-NOAA replied:
Done at bb7471f2. Thanks!

KateFriedman-NOAA committed:
Update jobs/JGDAS_FIT2OBS

Adjust comment for CDATE=${xdate}

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

KateFriedman-NOAA committed:
Update jobs/JGDAS_FIT2OBS

Add "FATAL ERROR" text to DATAROOT removal error message.

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

KateFriedman-NOAA committed:
Updates to jobs/JGDAS_FIT2OBS

- Revert double quotes to single quotes for COM_INF, need to pass
it in using single quotes, it can't be evaluated at this stage.
- Change CDUMP to RUN based on reviewer comments.

Refs #1405

WalterKolczynski-NOAA commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,82 @@
+export vcyc=$(echo "${xdate}" | cut -c9-10)
+
+export COM_INA=${ROTDIR}/gdas.${vday}/${vcyc}/atmos
+export COM_INF='$ROTDIR/vrfyarch/gfs.$fdy/$fzz'
```
```suggestion
# We want to defer variable expansion, so ignore warning about single quotes
# shellcheck disable=SC2016
export COM_INF='$ROTDIR/vrfyarch/gfs.$fdy/$fzz'
```

KateFriedman-NOAA committed:
Update jobs/JGDAS_FIT2OBS

Add shellcheck disable for SC2016 to ignore warning about single quotes.

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

WalterKolczynski-NOAA commented on jobs/JGDAS_FIT2OBS:
```diff
@@ -0,0 +1,84 @@
+
+export CDUMP=${RUN/enkf}
+
+export xdate=$(${NDATE} -"${VBACKUP_FITS}" "${CDATE}")
```
```suggestion
# Ignore spelling warning; nothing is misspelled
# shellcheck disable=SC2155
xdate=$(${NDATE} -"${VBACKUP_FITS}" "${PDY}${cyc}")
export xdate
```

KateFriedman-NOAA committed:
Update jobs/JGDAS_FIT2OBS

Update xdate to use PDY and cyc instead of CDATE.

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

KateFriedman-NOAA committed:
Resolve SC2155, SC2086, SC2248 and ignore SC2153

Refs #1405

KateFriedman-NOAA committed:
Remove use of xdate

- Remove xdate and just reset CDATE directly.

Refs #1405

KateFriedman-NOAA committed:
Remove exports of vday and vcyc; used locally

Refs #1405

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/fit2obs

KateFriedman-NOAA committed:
Resolve SC2086 and SC2292 for fit2obs section of config.resources

Refs #1405

aerorahul reviewed (approved):
looks good.
